### PR TITLE
osutil/mkfs: disable orphan_file feature for ext4

### PR DESCRIPTION
**Pull request by upils**
The SRU to fix [LP #2025339](https://bugs.launchpad.net/ubuntu/+source/e2fsprogs/+bug/2025339) has been rejected. So I would like to find a solution in snapd to fix https://bugs.launchpad.net/ubuntu-image/+bug/2028564.

Currently I disabled the `orphan_file` feature in every case but this will not work for series older than lunar because this feature was not present (mkfs.ext4 fails with a `Invalid filesystem option set: ^orphan_file` error). An ideal solution would be to do it only when running on Lunar. I am looking into it but do you know if there is already an easy way to check the series in snapd?

Currently, calling mkfs.ext4 without any option will result in a filesystem with the orphan_file feature disabled on mantic and up, so we do not loose anything setting it up explicitly for now.

### CODE REVIEW

**Comment by Meulengracht** [comment]
So just FTR, I believe the way to get the version of the host ubuntu is to utilize the information from `/etc/os-release`. This information is readily available in the following string: `release.ReleaseInfo.VersionID`, which contains for instance "22.04" or whatever version. You'd need to parse the string, or create a helper to do this.

**Comment by upils** [comment]
The current strategy is to check if we are running on a Lunar version. I saw we already do this kind of checks in other places in the code.

What we really would like to do is check if mkfs.ext4 supports the orphan_file feature and disable it if so. But this is more complex change, probably needing to either execute mkfs.ext4 a first time or/and properly parse `/etc/mk2fs.conf`. So let me know if you think that would be a cleaner solution.

File: osutil/mkfs/mkfs_test.go [hunk 1]
```diff
@@ -103,6 +104,23 @@ func (m *mkfsSuite) TestMkfsExt4Happy(c *C) {
 		},
 	})
 
+	cmd.ForgetCalls()
```
Comment (Meulengracht) [inline osutil/mkfs/mkfs_test.go:107]: These kinds of cases must be their own unit test, and not a part of a Happy case

Comment (upils) [reply osutil/mkfs/mkfs_test.go:107]: OK. The definition of the "Happy case" here is a bit blurry to me. I am used to consider "happy paths" as the ones not leading to errors but here I understand this is more like the "straight" path.

**Comment by alfonsosanchezbeato** [comment]
I think that the alternative would be to include `mkfs.ext4` in the `ubuntu-image` snap (this would solve other issues like a crash that happens when running in xenial). `ubuntu-image` could also ship all the different versions of `/etc/mke2fs.conf` from the different ubuntu releases (this files comes in `e2fsprogs` deb package).

According to the mkfs.ext4 man page, there is an env var called `MKE2FS_CONFIG` that can be used to determine the path of the configuration files. So `ubuntu-image` would do
1. Check UC version in the model assertion
3. Set `MKE2FS_CONFIG` accordingly (something like `$SNAP/<some_path>/<LTS_release>/mke2fs.conf`)
4. Then call snapd as usual
Doing things this way would make sure that `mkfs.ext4` uses always the default options for a given Ubuntu release, regardless of where it is running.

cc @sil2100

**Comment by upils** [comment]
@alfonsosanchezbeato thank you for the suggestion. I thought solving this in snapd would also prevent other use cases you may have from breaking. But if you think this is fine as is in snapd we can fix it in ubuntu-image.

Quick question regarding your suggestion: It means we would need to maintain a list of `mke2fs.conf` files, yes? I suspect this might be a problem because we would need to closely follow modifications made in the file when the `e2fsprogs` package is updated. Is there a recommended way on this?

**Comment by upils** [comment]
OK, thanks. I was already convinced adding e2fsprogs in the snap was needed, it will be 2 .

**Pull request closed without merging**